core/org.wso2.carbon.core/src/main/java/org/wso2/carbon/core/util/SignatureUtil.java

```diff
@@ -18,29 +18,45 @@
 package org.wso2.carbon.core.util;
 
 import org.wso2.carbon.base.MultitenantConstants;
+import org.wso2.carbon.base.ServerConfiguration;
 import org.wso2.carbon.base.api.ServerConfigurationService;
 import org.wso2.carbon.core.RegistryResources;
 import org.wso2.carbon.core.internal.CarbonCoreDataHolder;
+import org.wso2.carbon.utils.ServerConstants;
 
 import java.security.*;
 import java.security.cert.Certificate;
 import java.security.cert.X509Certificate;
 import java.util.Arrays;
 import java.util.Enumeration;
 
+import static org.wso2.carbon.core.util.CryptoUtil.getJCEProvider;
+
 public class SignatureUtil {
 
     private static final String THUMB_DIGEST_ALGORITHM = "SHA-1";
 
     private static String signatureAlgorithm = "SHA1withRSA";
-    private static String provider = "BC";
 
     private SignatureUtil() {
         // hide default constructor for utility class
     }
 
     public static void init() throws Exception {
-        Security.addProvider(new org.bouncycastle.jce.provider.BouncyCastleProvider());
+
+        String providerName = ServerConfiguration.getInstance()
+                .getFirstProperty(ServerConstants.JCE_PROVIDER);
+        Provider provider;
+        if (providerName.equals(ServerConstants.JCE_PROVIDER_BCFIPS)) {
+            provider = (Provider) (Class.forName("org.bouncycastle.jcajce.provider.BouncyCastleFipsProvider")).newInstance();
+
+        } else if (providerName.equals(ServerConstants.JCE_PROVIDER_BC)) {
+            provider = (Provider) (Class.forName("org.bouncycastle.jce.provider.BouncyCastleProvider")).newInstance();
+
+        } else {
+            throw new CryptoException("Invalid JCE provider");
```

Comment:
Can we use `NotSupportedException`? Also let's improve the error message `Configured JCE provider is not supported.`

Reply:
Updated with NoSuchProviderException